provider/adapters/src/fallback/adapter.rs

```diff
@@ -3,25 +3,186 @@
+///     .map(|_| ())
+///     .expect_err("No data for ar-SA");
+///
+/// // But if we wrap the provider in a fallback provider...
+/// let provider = LocaleFallbackProvider::try_new(provider)
+///     .expect("Fallback data present");
+///
+/// // ...then we can load "ar-SA" based on "ar" data
+/// let pr = PluralRules::try_new_cardinal(locale!("ar-SA"), &provider)
+///     .expect("Using fallback provider");
+///
+/// assert_eq!(pr.select(5_usize), icu_plurals::PluralCategory::Few);
+/// ```
+pub struct LocaleFallbackProvider<P> {
+    pub inner: P,
+    fallbacker: LocaleFallbacker,
+}
+
+impl<P> LocaleFallbackProvider<P>
+where
+    P: ResourceProvider<LocaleFallbackLikelySubtagsV1Marker>
+        + ResourceProvider<LocaleFallbackParentsV1Marker>,
+{
+    /// Create a [`LocaleFallbackProvider`] by wrapping another data provider and then loading
+    /// fallback data from it.
+    ///
+    /// If the data provider being wrapped does not contain fallback data, use
+    /// [`LocaleFallbackProvider::new_with_fallbacker`].
+    pub fn try_new(provider: P) -> Result<Self, DataError> {
+        let fallbacker = LocaleFallbacker::try_new(&provider)?;
+        Ok(Self {
+            inner: provider,
+            fallbacker,
+        })
+    }
 }
 
-pub struct LocaleFallbackAdapter<P>(pub P);
+impl<P> LocaleFallbackProvider<P> {
+    /// Wrap a provider with an arbitrary fallback engine.
+    ///
+    /// This relaxes the requirement that the wrapped provider contains its own fallback data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_provider::hello_world::*;
+    /// use icu_provider::prelude::*;
+    /// use icu_provider_adapters::fallback::{LocaleFallbacker, LocaleFallbackProvider};
+    ///
+    /// let provider = HelloWorldProvider::new_with_placeholder_data();
+    ///
+    /// let req = DataRequest {
+    ///     options: locale!("de-CH").into(),
+    ///     metadata: Default::default(),
+    /// };
+    ///
+    /// // There is no "de-CH" data in the `HelloWorldProvider`
+    /// ResourceProvider::<HelloWorldV1Marker>::load_resource(&provider, &req)
+    ///     .expect_err("No data for de-CH");
+    ///
+    /// // `HelloWorldProvider` does not contain fallback data,
+    /// // but we can fetch it from `icu_testdata`, and then
+    /// // use it to create the fallbacking data provider.
+    /// let fallbacker = LocaleFallbacker::try_new(&icu_testdata::get_provider())
+    ///     .expect("Fallback data present");
+    /// let provider = LocaleFallbackProvider::new_with_fallbacker(provider, fallbacker);
+    ///
+    /// // Now we can load the "de-CH" data via fallback to "de".
+    /// let german_hello_world: DataPayload<HelloWorldV1Marker> = provider
+    ///     .load_resource(&req)
+    ///     .expect("Loading should succeed")
+    ///     .take_payload()
+    ///     .expect("Data should be present");
+    ///
+    /// assert_eq!("Hallo Welt", german_hello_world.get().message);
```

Comment:
Yoda condition

Reply:
I take this comment to mean that you would prefer that I write "actual, expected" rather than "expected, actual"?

I find "expected, actual" to be more readable in this case because you have a short literal (the expected value) followed by a long expression (the actual value). From Googling around a bit, it doesn't appear that there is a very strong consensus. So I prefer to keep it with expected value first, but I don't know if you feel more strongly about the matter than I.